APAGovernance.sol

```diff
@@ -0,0 +1,169 @@
+    uint public proposerApas; 
+    uint private nextPropId;
+    Proposal[] public proposals;
+    mapping(uint => mapping(address => bool)) public voters;
+    mapping(uint => mapping(uint => bool)) public votedAPAs;
+    address private apaToken = 0x9E491465BbD22B62D7d27E0Ff35d4e263228ba5C;
+    address private apaMarket = 0x451650527d1a58875E71cf18F429E46838387Dc1;
+
+
+    constructor() {
+        manager = msg.sender;
+        apaContract = IERC721Enumerable(apaToken);
+        proposerApas = 30;
+    }
+
+    modifier onlyManager() {
+        require(msg.sender == manager, 'only manager can execute this function');
+        _;
+    } 
+
+    modifier verifyNumApas(uint minAPAs) {
+        // Check if sender has minimum number of APAs
+        require(balanceOf(msg.sender) >= minAPAs, 'Need more APAs');
+        _;
+    }
+    //function to set number of APAs needed to be able to create proposal
+    function setProposerApas(uint minApas) public onlyManager() {
+        require (minApas >0, "set minimum to at least one APA ");
+        proposerApas = minApas;
+    }
+
+    function balanceOf(address owner) 
+        public
+        view 
+        returns (uint){
+        return apaContract.balanceOf(owner);
+    }
+
+    function createProposal(
+        string memory _name, 
+        string memory _desc,
+        uint duration, //in days
+        BallotType _ballotType //0=perAPA 1=perAddress
+        ) external verifyNumApas(proposerApas)  {
+            proposals.push( Proposal({
+                id: nextPropId,
+                author: msg.sender,
+                name: _name,
+                description: _desc,
+                end: block.timestamp + duration * 1 days,
+                votesYes: 0,
+                votesNo: 0,
+                ballotType: _ballotType,
+                status: Status.Active
+            })
+            );
+            nextPropId+=1;
+
+    }
+
+    function vote(uint proposalId , Votes _vote) external {
+        uint voterBalance = balanceOf(msg.sender);
+        require(voterBalance > 0, "Need at least one APA to cast a vote");
+        require(!voters[proposalId][msg.sender], "Voter has already voted");  // Maker sure voter has not yet voted
+        require(block.timestamp < proposals[proposalId].end, "Proposal has Expired");
+        uint currentAPA;
+        uint ineligibleCount=0;
+ 
+        //get user Apas map from wallet
+       for(uint i=0; i < voterBalance; i++){
+            currentAPA = apaContract.tokenOfOwnerByIndex(msg.sender, i);
+            //check if APA has already voted and add to UserAPAs if eligible
+           
+            if(votedAPAs[proposalId][currentAPA] != true){
+                //add APAs to proposal array
+                if(proposals[proposalId].ballotType == BallotType.perAPA){
+                    votedAPAs[proposalId][currentAPA] = true;
+                }
+                if(proposals[proposalId].ballotType == BallotType.perAddress){
```

Comment:
do we need to iterate through whole array if the voting is limited to one per address ?

Reply:
does the break statement on line 118 take care of this? it should break out of for loop when it finds one eligible apa